### DISCUSSION TITLE
File Handling Questions

### DISCUSSION THREAD

**Discussion by maxwkfcc** [duncanmcclean/guest-entries#37] [Help] [answered]
Inside the Guest Entries, while I am updating files in entry, it is replacing my original image.  It is reasonable for a single file upload.

1. However, while I am using multiple image upload, and I add a new image in the upload box and save, it will remove all my images and replaced will all new images.  How should I prevent this?

2. And, how should I remove a single image?

3. Another one I can think of is, after form validation via FormRequest, how should I persist the file in the input box or how to let user not to input those files again?

https://discord.com/channels/489818810157891584/489862989101662218/1063423057923944459

**Comment by duncanmcclean** [comment, marked as answer]
Hey! Thanks for creating this discussion following our discussion on Discord.

> 1. However, while I am using multiple image upload, and I add a new image in the upload box and save, it will remove all my images and replaced will all new images. How should I prevent this?

I don't think this is currently possible without asking the user to re-upload each of the images. 

However, I could make it possible to reference the filenames of existing assets so you can keep hold of them. As an example, it might look something like this:

```antlers
{{ images }}
    <input type="hidden" name="images[]" value="{{ path }}">
{{ /images }}

<input type="file" name="images[]">
```

You'd loop through any of the assets already uploaded to the `images` field, then you'd have an input to allow the user to upload more images.

> 2. And, how should I remove a single image?

This is another one that currently isn't possible. However, once I've implemented the feature I was talking about in 1, you would simply delete the `<input>`  field for the image you wish to delete. This would probably be something you'd need to add a little bit of JavaScript to achieve.

> 3. Another one I can think of is, after form validation via FormRequest, how should I persist the file in the input box or how to let user not to input those files again?

Apologies, I'm not sure I quite understand what you mean. Are you asking how the files can be kept in the form when the user goes to edit the same entry. If so, I've covered that in 1.

**Reply by maxwkfcc** [reply]
Hi Duncan,

Thanks for your reply.  I should state clearer for Q3.

While submission of a form, if there are some fields cannot pass the validation condition, the same form will be reloaded with filled old values and error messages.  So, for the image/file fields, the old value cannot be populated, any trick to achieve this?

Thanks
Max

**Reply by duncanmcclean** [reply]
Not that I know of, sorry!

There is a [StackOverflow article](https://stackoverflow.com/questions/33158328/laravel-5-1-keep-uploaded-file-as-old-input) talking about it - unsure if it's of any help but it's all I could find.

**Reply by maxwkfcc** [reply]
Hi Duncan,

I tried to extended your controller and update the `public function update(UpdateRequest $request)` to add images to the list.

```
foreach (Arr::except($request->all(), $this->ignoredParameters) as $key => $value) {
            /** @var \Statamic\Fields\Field $blueprintField */
            $field = $entry->blueprint()->field($key);

            /** This HACK is to add image to existing image rather than replacing them */
            $processedData = $field
                ? $this->processField($field, $key, $value, $request)
                : $value;

            // Check if AssetFieldtype
            if ($field && $field->fieldtype() instanceof AssetFieldtype) {
                
                // Check if allow multiple images
                // TODO: need to get the field settings to see if the upload limit is > 1 or not
                // if ($isAllowMultipleImages) {
                    if (!$processedData) $processedData = [];
                    if (!is_array($processedData)) $processedData = [$processedData];
                    $processedData = array_merge($data[$key], $processedData);
                // }
            }



            $data[$key] = $processedData;
        }
```

However, I need to edit the actions.php inside your package to
```
Route::name('guest-entries.')->group(function () {
    Route::post('/create', [GuestEntryMyController::class, 'store'])->name('store');
    Route::post('/update', [GuestEntryMyController::class, 'update'])->name('update');
    Route::delete('/delete', [GuestEntryMyController::class, 'destroy'])->name('destroy');
});
```

Do you know any better way to do this hack like without editing the actions.php?

Thanks
Max

**Reply by duncanmcclean** [reply]
You can create your own route in your `routes/web.php` file. 

Then, instead of using the provided Antlers Tag to make the form, construct the `<form>` manually and update the `action` of the form to point at your controller.

```html
<form method="POST" action="/your-custom-route">
    {{ csrf_field }}
    <input type="hidden" name="_collection" value="COLLECTION HANDLE GOES HERE" />
    <input type="hidden" name="_id" value="ENTRY ID GOES HERE" />

    <!-- Then, everything you had inside the Guest Entries tag before, put here. -->
</form>
```

Does that help?

**Reply by maxwkfcc** [reply]
It works for me.  Thanks Duncan
